### ISSUE TITLE
Numpy FFT type promotion -> complex128

### ISSUE THREAD

**Issue by timlod** [librosa/librosa#1074]
#### Description
By default, librosa uses the numpy fft implementation (numpy.fft) when calling librosa.get_fftlib().
That implementation will always promote input from float32 such that the output is complex128 (2 64bit floats to make the complex number) [(Numpy fft type promotion)](https://numpy.org/devdocs/reference/routines.fft.html#type-promotion).
This can silently double memory consumption in pipelines that use methods based on fft.

librosa has `scipy` as a requirement, so it would be possible to default to `scipy.fft` instead, which respects the input types.

Currently this can be achieved by doing this before running any memory-constrained code:
```
import scipy.fft as scipyfft
librosa.set_fftlib(scipyfft)
```

#### Steps/Code to Reproduce

```
import librosa

S = np.float32(np.random.randn(1025))
S_cqt = librosa.cqt(S) # (uses fft internally)
print(S_cqt.dtype) # np.complex128
```

#### Expected Results
`print(S_cqt.dtype) # np.complex64`

#### Actual Results
`print(S_cqt.dtype) # np.complex128`

#### Versions
Librosa 0.7.2

**Comment by bmcfee** [comment]
This is a fair point, thanks for raising it.

If I understand correctly, your suggestion is to make fftpack the default again?  (It was previously, but we reverted to numpy's implementation after doing some speed benchmarking.)

I'm not opposed to this, but it does warrant some discussion.

**Comment by timlod** [comment]
I've not been using librosa extensively, so I don't know all the arguments for or against using either - but I did notice this particular issue and wanted to raise it, as Python carries on quietly without users realising certain behaviours. If you value consistency in types/memory consumption over some possible tradeoffs with regards to execution time, I would suggest it. In any case, I would add the fact that anything will be cast to 64bit resolution when using numpy's fft methods.

**Comment by lostanlen** [comment]
Thank you for raising this, @timlod.
Related: #825 (closed by #832)

**Comment by bmcfee** [comment]
Gave this a think, and here are some ideas:

1. STFT already has a `dtype` parameter to force expected behavior, regardless of the underlying FFT implementation.  (It could be smarter though, inferring the dtype from the precision of the input.)
2. OP's example code gets type promotion through CQT, not STFT.  This could be fixed by adding a similar `dtype=` parameter to CQT methods, and I'd support that in general.
3. One additional thing I've been pondering is extending the STFT API to allow the user to pre-allocate an output array, much like numpy ufuncs.  This would mainly benefit iterative procedures like griffin-lim or CQT (which both use temporary stft arrays).

We are ultimately limited in how far we can push the memory efficiency efforts by virtue of abstracting the FFT backend.  Since the FFT api does not allow a pre-allocated array, it will always allocate its own buffers, which will then be copied (and cast, if necessary) into our allocated buffer.  Working in small blocks (as we currently do) mitigates this somewhat, but it's not an entirely ideal solution.
